core/src/main/java/com/linecorp/armeria/client/RedirectingClient.java

```diff
@@ -580,17 +576,64 @@ String originalUri() {
             return originalUri;
         }
 
-        boolean addRedirectUri(HttpMethod method, String redirectUri) {
-            if (redirectUris == null) {
-                redirectUris = LinkedListMultimap.create();
+        boolean addRedirectSignature(String redirectUri, HttpMethod method) {
+            if (redirectSignatures == null) {
+                redirectSignatures = ConcurrentHashMap.newKeySet();
+            }
+
+            if (!isAddedOriginalSignature) {
+                final RedirectSignature originalSignature =
+                        new RedirectSignature(originalUri(), request.method());
+                redirectSignatures.add(originalSignature);
+                isAddedOriginalSignature = true;
             }
-            return redirectUris.put(method, redirectUri);
+
+            final RedirectSignature signature = new RedirectSignature(redirectUri, method);
+            return redirectSignatures.add(signature);
+        }
+
+        Set<RedirectSignature> redirectSignatures() {
+            // Always called after addRedirectSignature is called.
+            assert redirectSignatures != null;
+            return redirectSignatures;
+        }
+
+        Set<String> redirectUris() {
+            // Always called after addRedirectSignature is called.
+            assert redirectSignatures != null;
+            return redirectSignatures.stream()
+                                     .map(signature -> signature.uri)
+                                     .collect(ImmutableSet.toImmutableSet());
         }
+    }
+
+    @VisibleForTesting
+    static class RedirectSignature {
+        private final String uri;
```

Comment:
I think we need to decompose this into:
- whether it's TLS or not,
- authority and path.

Because a redirect URI can contain schemes like `h2` or `h2c` that should be treated as `HTTPS` or `HTTP` (`h1` and `h1c` as well).

The current code calls `nextTarget.toString()` to generate the full URI, but we should instead retrieve the individual components from `nextTarget` to generate the correct signature.

Once fixed, we could also add the test cases for this as well.

Reply:
> Because a redirect URI can contain schemes like h2 or h2c that should be treated as HTTPS or HTTP (h1 and h1c as well).
The current code calls nextTarget.toString() to generate the full URI, but we should instead retrieve the individual components from nextTarget to generate the correct signature.

I think the URLs has already been converted, such as `h1c`, `h2c` -> `HTTP` before, adding them as signature:
* originalURL: [buildUri()](https://github.com/line/armeria/blob/db3973d74e70e177b0849f201fdd96615f2f6292/core/src/main/java/com/linecorp/armeria/client/RedirectingClient.java#L483)
* redirectURLs: [resolveLocation()](https://github.com/line/armeria/blob/db3973d74e70e177b0849f201fdd96615f2f6292/core/src/main/java/com/linecorp/armeria/client/RedirectingClient.java#L295)

So, I used the URLs directly and added the following tests:
https://github.com/line/armeria/blob/40b949c22c9ac72c405939e46bb7faf6ef88a9e4/core/src/test/java/com/linecorp/armeria/client/RedirectingClientTest.java#L127-L148

Could you check this as it may be an omission in my consideration?

Reply:
Ahh, I see. But still, we're doing too much string manipulation in `RequestTarget.toString()` and `buildUri()`. Could we avoid most string concatenations we do here, at least until we construct the exception? This way, we could reduce GC footprint for normal cases.

We could use the following properties as a part of redirect signature.

- `ctx.sessionProtocol().isTls() ? "https" : "http"` / `nextTarget.scheme()`
- `ctx.authority()` / `nextTarget.authority()`
- `headers.path()` / `nextTarget.pathAndQuery()`

I assumed the authority returned by `ctx.authority()` is normalized by default, because `RequestTarget` always normalized it. If this is not the case, we can handle it as a separate issue.

Reply:
With `ctx.authority()` and others, I could keep string concatenation only within Exception.

However, there are a few cases where this does not work quite right.
If no port is specified in authority, `buildUri()` adds a port to authority depending on protocol, endpoint, but `ctx.authority()` does not include a port.

e.g.,. 
Consider the following redirect
```
1 -> 2 -> 3
1. http://foo
2. http://foo:80
3. http://foo or http://foo:80
```

I think there will be one extra redirect, although the circulation can be detected, because of the following differences in the handling of the original URI.
* `signature.uri()`: http://foo
* `buildUri()`: http://foo:80

So, it might be better to handle adding it only if `ctx.authority()` does not contain a port (maybe separate from this Issue?).
If this could be done, I think it would also be possible to remove `buildUri()`.

Reply:
Could we assume `ctx.authority()` and `RequestTarget.authority()` always returns an optimal form, omitting redundant port numbers? Then we'll never see `http://foo:80` nor `https://foo:443`.

At least for now, let's pretend it is so, because the problem will probably go away once #5490 is resolved.